functorch/experimental/_cond.py

```diff
@@ -149,7 +157,66 @@ def cond_python_dispatcher(*args):
     return cond(*args)
 
 
+def _has_potential_branch_input_mutation(branch, fake_inputs):
+    """
+    Dispatch-trace the branch with fake inputs and check if
+    producing graph has mutable op on the input. This is
+    bit restrictive as the branch must be traceable.
+    """
+    try:
+        gm = make_fx(branch)(*fake_inputs)
+    except UnsupportedMutationException:
+        # this can happen when nested cond is
+        # functionalized
+        return True
+    except Exception as e:
+        raise e
+
+    input_nodes = set()
+    for node in gm.graph.nodes:
+        if node.op == "placeholder":
+            input_nodes.add(node)
+
+    for node in gm.graph.nodes:
+        if node.op == "call_function":
+            target = node.target
+            if isinstance(target, torch._ops.OpOverload) and target._schema.is_mutable:
+                if node.args[0] in input_nodes:
+                    return True
+
+    return False
```

Comment:
@bdhirsh wrote somewhere:

> One thing I'm not totally convinced of: does this work if your true/false branch returns outputs that alias inputs, or if they mutate inputs?
> It might be easiest just to assert that neither of those cases happen for now, and add that support later.

If we want to be really safe, then we should also assert that the inputs are not aliased. The problematic case is if a branch aliases an input and then mutates the alias (this ends up mutating the input, but our check does not detect it).

I'm not sure if the schema (target._schema) in Python gives enough information for us to check this though, does it?

Reply:
Agreed with Richard that the input mutation check isn't comprehensive enough. We actually do similar checks in a few other places in the codebase, e.g. [here](https://github.com/pytorch/pytorch/blob/01e7f462152ec0bd34839c4a5513ccef4eb12b8a/torch/fx/passes/reinplace.py#L536).

Since you're tracing through the graph with fake tensors, you have all the info on aliasing that you need. You can do something like this to track the storages of all input tensors, and then check if a given mutated node shares a storage with any inputs:
```
    input_storages = set(
        StorageWeakRef(
            node.meta['val']._typed_storage()
        ) for node in gm.graph.nodes if node.op == 'placeholder')
```

Reply:
added